File: android/Reguerta/app/src/main/java/com/reguerta/user/data/profiles/ChainedSharedProfileRepository.kt
package com.reguerta.user.data.profiles

import com.reguerta.user.domain.profiles.SharedProfile
import com.reguerta.user.domain.profiles.SharedProfileRepository

class ChainedSharedProfileRepository(
    private val primary: SharedProfileRepository,
    private val fallback: SharedProfileRepository,
) : SharedProfileRepository {
    override suspend fun getAllSharedProfiles(): List<SharedProfile> {
        val primaryProfiles = primary.getAllSharedProfiles()
        return if (primaryProfiles.isNotEmpty()) {
            primaryProfiles
        } else {
            fallback.getAllSharedProfiles()
        }
    }

    override suspend fun getSharedProfile(userId: String): SharedProfile? =
        primary.getSharedProfile(userId) ?: fallback.getSharedProfile(userId)

    override suspend fun upsertSharedProfile(profile: SharedProfile): SharedProfile =
        primary.upsertSharedProfile(profile)

    override suspend fun deleteSharedProfile(userId: String): Boolean =
        primary.deleteSharedProfile(userId)
}

File: ios/Reguerta/Reguerta/Data/Profiles/ChainedSharedProfileRepository.swift
import Foundation

final class ChainedSharedProfileRepository<Primary: SharedProfileRepository, Fallback: SharedProfileRepository>:
    @unchecked Sendable, SharedProfileRepository {
    private let primary: Primary
    private let fallback: Fallback

    init(primary: Primary, fallback: Fallback) {
        self.primary = primary
        self.fallback = fallback
    }

    func allSharedProfiles() async throws -> [SharedProfile] {
        let primaryProfiles = try await primary.allSharedProfiles()
        return primaryProfiles.isEmpty ? try await fallback.allSharedProfiles() : primaryProfiles
    }

    func sharedProfile(userId: String) async throws -> SharedProfile? {
        if let profile = try await primary.sharedProfile(userId: userId) {
            return profile
        }
        return try await fallback.sharedProfile(userId: userId)
    }

    func upsert(profile: SharedProfile) async throws -> SharedProfile {
        try await primary.upsert(profile: profile)
    }

    func deleteSharedProfile(userId: String) async throws -> Bool {
        try await primary.deleteSharedProfile(userId: userId)
    }
}
